File: rust/parser/src/inlines.rs
//! The inline algorithm (SPEC.md, "The inline algorithm"): one left-to-right
//! pass per container; escape > code span > bracket > emoji > delimiter run;
//! top-of-stack delimiter matching; everything unmatched reverts to literal.

use crate::ast::{Attrs, Node};
use crate::attrs::{is_name, name_len, parse_attrs, parse_value};
use crate::blocks::MAX_TARGET_BYTES;

pub const MAX_INLINE_DEPTH: usize = 8;
pub const MAX_EMOJI_SLUG_BYTES: usize = 64;

#[derive(Clone, Copy, PartialEq, Eq)]
enum DelimKind {
    Em,
    Strong,
    Strike,
}

struct Delim {
    kind: DelimKind,
    idx: usize,
    raw: &'static str,
}

struct Frame {
    /// Raw span opener text (`[color=red]`), for reverting if never closed.
    opener_raw: String,
    name: String,
    attrs: Attrs,
    children: Vec<Node>,
    delims: Vec<Delim>,
}

impl Frame {
    fn root() -> Frame {
        Frame {
            opener_raw: String::new(),
            name: String::new(),
            attrs: Attrs::new(),
            children: Vec::new(),
            delims: Vec::new(),
        }
    }
}

pub fn parse_inlines(text: &str) -> Vec<Node> {
    let chars: Vec<char> = text.chars().collect();
    let mut frames: Vec<Frame> = vec![Frame::root()];
    let mut i = 0;

    while i < chars.len() {
        let c = chars[i];
        match c {
            '\\' => match chars.get(i + 1) {
                Some('\n') => {
                    frames.last_mut().unwrap().children.push(Node::HardBreak);
                    i += 2;
                }
                Some(&n) if n.is_ascii() && n.is_ascii_punctuation() => {
                    push_char(&mut frames, n);
                    i += 2;
                }
                _ => {
                    push_char(&mut frames, '\\');
                    i += 1;
                }
            },
            '`' => {
                let n = run_len(&chars, i, '`');
                match find_backtick_closer(&chars, i + n, n) {
                    Some(k) => {
                        let text: String = chars[i + n..k].iter().collect();
                        frames.last_mut().unwrap().children.push(Node::CodeSpan { text });
                        i = k + n;
                    }
                    None => {
                        push_str(&mut frames, &"`".repeat(n));
                        i += n;
                    }
                }
            }
            '[' => i = bracket(&chars, i, false, &mut frames),
            '!' if chars.get(i + 1) == Some(&'[') => i = bracket(&chars, i + 1, true, &mut frames),
            ':' => {
                let mut k = i + 1;
                while k < chars.len() && is_slug_char(chars[k]) {
                    k += 1;
                }
                let slug_len = k - (i + 1);
                if (1..=MAX_EMOJI_SLUG_BYTES).contains(&slug_len)
                    && chars.get(k) == Some(&':')
                {
                    let slug: String = chars[i + 1..k].iter().collect();
                    frames.last_mut().unwrap().children.push(Node::Emoji { slug });
                    i = k + 1;
                } else {
                    push_char(&mut frames, ':');
                    i += 1;
                }
            }
            '*' => {
                let n = run_len(&chars, i, '*');
                match n {
                    1 => i = delimiter(&chars, i, n, DelimKind::Em, "*", &mut frames),
                    2 => i = delimiter(&chars, i, n, DelimKind::Strong, "**", &mut frames),
                    _ => {
                        push_str(&mut frames, &"*".repeat(n));
                        i += n;
                    }
                }
            }
            '~' => {
                let n = run_len(&chars, i, '~');
                if n == 2 {
                    i = delimiter(&chars, i, n, DelimKind::Strike, "~~", &mut frames);
                } else {
                    push_str(&mut frames, &"~".repeat(n));
                    i += n;
                }
            }
            c => {
                push_char(&mut frames, c);
                i += 1;
            }
        }
    }

    // Container end: unclosed spans and delimiters revert to literal text.
    while frames.len() > 1 {
        let frame = frames.pop().unwrap();
        let flat = revert_frame(frame);
        let parent = frames.last_mut().unwrap();
        parent.children.extend(flat);
    }
    let root = frames.pop().unwrap();
    normalize(revert_delims(root.children, root.delims))
}

fn push_char(frames: &mut [Frame], c: char) {
    let mut buf = [0u8; 4];
    push_str(frames, c.encode_utf8(&mut buf));
}

fn push_str(frames: &mut [Frame], s: &str) {
    let frame = frames.last_mut().unwrap();
    // An open delimiter sits (invisibly, until it closes) at its recorded
    // index: text on its far side must not merge into text before it.
    let barrier = frame.delims.last().map_or(0, |d| d.idx);
    let children = &mut frame.children;
    if children.len() > barrier {
        if let Some(Node::Text { value }) = children.last_mut() {
            value.push_str(s);
            return;
        }
    }
    children.push(Node::Text { value: s.to_string() });
}

fn run_len(chars: &[char], i: usize, c: char) -> usize {
    chars[i..].iter().take_while(|&&x| x == c).count()
}

/// Grammar whitespace is ASCII only (SPEC.md, front-door normalization):
/// Unicode spaces are content, not structure.
fn is_ws(c: char) -> bool {
    matches!(c, ' ' | '\t' | '\n')
}

fn is_slug_char(c: char) -> bool {
    c.is_ascii_lowercase() || c.is_ascii_digit() || matches!(c, '_' | '+' | '-')
}

fn find_backtick_closer(chars: &[char], from: usize, n: usize) -> Option<usize> {
    let mut k = from;
    while k < chars.len() {
        if chars[k] == '`' {
            let m = run_len(chars, k, '`');
            if m == n {
                return Some(k);
            }
            k += m;
        } else {
            k += 1;
        }
    }
    None
}

fn total_depth(frames: &[Frame]) -> usize {
    frames.len() - 1 + frames.iter().map(|f| f.delims.len()).sum::<usize>()
}

fn delimiter(
    chars: &[char],
    i: usize,
    n: usize,
    kind: DelimKind,
    raw: &'static str,
    frames: &mut [Frame],
) -> usize {
    let can_close = i > 0 && !is_ws(chars[i - 1]);
    let can_open = chars.get(i + n).is_some_and(|&c| !is_ws(c));
    let deep = total_depth(frames) >= MAX_INLINE_DEPTH;
    let frame = frames.last_mut().unwrap();
    if can_close && frame.delims.last().is_some_and(|d| d.kind == kind) {
        let delim = frame.delims.pop().unwrap();
        let inner: Vec<Node> = frame.children.drain(delim.idx..).collect();
        let inner = normalize(inner);
        frame.children.push(match kind {
            DelimKind::Em => Node::Emphasis { children: inner },
            DelimKind::Strong => Node::Strong { children: inner },
            DelimKind::Strike => Node::Strikethrough { children: inner },
        });
    } else if can_open && !deep {
        let idx = frame.children.len();
        frame.delims.push(Delim { kind, idx, raw });
    } else {
        push_str(frames, raw);
    }
    i + n
}

/// Handle a bracket construct starting at `chars[open]` (which is `[`).
/// `embed` means a `!` sits just before it. Returns the new position.
fn bracket(chars: &[char], open: usize, embed: bool, frames: &mut Vec<Frame>) -> usize {
    let bang = if embed { open - 1 } else { open };
    let fallback = |frames: &mut Vec<Frame>| {
        push_char(frames, chars[bang]);
        bang + 1
    };

    // Find the matching `]` (balanced, escape-aware).
    let mut depth = 1usize;
    let mut k = open + 1;
    while k < chars.len() {
        match chars[k] {
            '\\' => k += 1,
            '[' => depth += 1,
            ']' => {
                depth -= 1;
                if depth == 0 {
                    break;
                }
            }
            _ => {}
        }
        k += 1;
    }
    if k >= chars.len() {
        return fallback(frames);
    }
    let interior: String = chars[open + 1..k].iter().collect();

    // Link / embed: `](` with a lexable target.
    if chars.get(k + 1) == Some(&'(') {
        if let Some((target, end)) = lex_target(chars, k + 2) {
            let node = if embed {
                Node::Embed { target, alt: resolve_escapes(&interior) }
            } else {
                Node::Link { target, children: parse_inlines(&interior) }
            };
            frames.last_mut().unwrap().children.push(node);
            return end;
        }
        return fallback(frames);
    }

    // Span closer: `[/name]` must name the innermost open span.
    if let Some(name) = interior.strip_prefix('/') {
        if is_name(name) {
            if frames.len() > 1 && frames.last().unwrap().name == name {
                let frame = frames.pop().unwrap();
                let children = normalize(revert_delims(frame.children, frame.delims));
                frames.last_mut().unwrap().children.push(Node::Span {
                    name: frame.name,
                    attrs: frame.attrs,
                    children,
                });
            } else {
                // Well-formed but mismatched or orphan: the characters back.
                push_str(frames, &format!("[/{name}]"));
            }
            return k + 1;
        }
        return fallback(frames);
    }

    // Span opener: `[name ...]`, with the BBCode default-parameter idiom
    // (`[color=red]` puts `color=red` in the span's own attrs).
    if let Some((name, attrs)) = parse_span_opener(&interior) {
        if total_depth(frames) >= MAX_INLINE_DEPTH {
            let raw: String = chars[open..=k].iter().collect();
            push_str(frames, &raw);
        } else {
            frames.push(Frame {
                opener_raw: chars[open..=k].iter().collect(),
                name,
                attrs,
                children: Vec::new(),
                delims: Vec::new(),
            });
        }
        return k + 1;
    }

    fallback(frames)
}

fn parse_span_opener(interior: &str) -> Option<(String, Attrs)> {
    let nlen = name_len(interior);
    if nlen == 0 {
        return None;
    }
    let name = &interior[..nlen];
    let mut attrs = Attrs::new();
    let mut rest = &interior[nlen..];
    if rest.starts_with('=') {
        let (value, after) = parse_value(&rest[1..]).ok()?;
        if !(after.is_empty() || after.starts_with(' ') || after.starts_with('\t')) {
            return None;
        }
        attrs.insert(name.to_string(), value);
        rest = after;
    } else if !(rest.is_empty() || rest.starts_with(' ') || rest.starts_with('\t')) {
        return None;
    }
    for (key, value) in parse_attrs(rest).ok()? {
        attrs.entry(key).or_insert(value);
    }
    Some((name.to_string(), attrs))
}

/// Lex a link/embed target from `chars[from]`: no whitespace, balanced
/// parens, an unbalanced `)` ends it. Returns (target, position after `)`).
fn lex_target(chars: &[char], from: usize) -> Option<(String, usize)> {
    let mut depth = 0usize;
    let mut k = from;
    while k < chars.len() {
        match chars[k] {
            ')' if depth == 0 => {
                let target: String = chars[from..k].iter().collect();
                if target.len() > MAX_TARGET_BYTES {
                    return None;
                }
                return Some((target, k + 1));
            }
            ')' => depth -= 1,
            '(' => depth += 1,
            c if is_ws(c) => return None,
            _ => {}
        }
        k += 1;
    }
    None
}

fn resolve_escapes(s: &str) -> String {
    let mut out = String::new();
    let mut chars = s.chars();
    while let Some(c) = chars.next() {
        if c == '\\' {
            match chars.next() {
                Some(n) if n.is_ascii() && n.is_ascii_punctuation() => out.push(n),
                Some(n) => {
                    out.push('\\');
                    out.push(n);
                }
                None => out.push('\\'),
            }
        } else {
            out.push(c);
        }
    }
    out
}

/// An unclosed span reverts: its opener text, then its children, flattened
/// into the parent (its own unmatched delimiters reverting first).
fn revert_frame(frame: Frame) -> Vec<Node> {
    let mut out = vec![Node::Text { value: frame.opener_raw }];
    out.extend(revert_delims(frame.children, frame.delims));
    out
}

fn revert_delims(mut children: Vec<Node>, delims: Vec<Delim>) -> Vec<Node> {
    for d in delims.into_iter().rev() {
        children.insert(d.idx, Node::Text { value: d.raw.to_string() });
    }
    children
}

/// Canonical text: adjacent literals merge, empty text nodes vanish.
fn normalize(children: Vec<Node>) -> Vec<Node> {
    let mut out: Vec<Node> = Vec::new();
    for node in children {
        match node {
            Node::Text { value } if value.is_empty() => {}
            Node::Text { value } => {
                if let Some(Node::Text { value: prev }) = out.last_mut() {
                    prev.push_str(&value);
                } else {
                    out.push(Node::Text { value });
                }
            }
            other => out.push(other),
        }
    }
    out
}

File: ts/parser/src/inlines.ts
// The inline algorithm (SPEC.md, "The inline algorithm"): a faithful port of
// rust/parser/src/inlines.rs. One left-to-right pass per container; escape >
// code span > bracket > emoji > delimiter run; top-of-stack delimiter
// matching; everything unmatched reverts to literal.
//
// Works on code points (Array.from), mirroring Rust's Vec<char>, so the two
// implementations scan identical units.

import type { Attrs, Node } from "./ast.ts";
import { isName, nameLen, parseAttrs, parseValue, utf8Len } from "./attrs.ts";
import { MAX_TARGET_BYTES } from "./blocks.ts";

export const MAX_INLINE_DEPTH = 16;
export const MAX_EMOJI_SLUG_BYTES = 64;

type DelimKind = "em" | "strong" | "strike";

interface Delim {
  kind: DelimKind;
  idx: number;
  raw: string;
}

interface Frame {
  /** Raw span opener text (`[color=red]`), for reverting if never closed. */
  openerRaw: string;
  name: string;
  attrs: Attrs;
  children: Node[];
  delims: Delim[];
}

function rootFrame(): Frame {
  return { openerRaw: "", name: "", attrs: {}, children: [], delims: [] };
}

const ASCII_PUNCT = new Set("!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~");

/** Grammar whitespace is ASCII only (SPEC.md, front-door normalization):
 * Unicode spaces are content, not structure. */
function isWs(c: string): boolean {
  return c === " " || c === "\t" || c === "\n";
}

function isSlugChar(c: string): boolean {
  return /^[a-z0-9_+-]$/.test(c);
}

export function parseInlines(text: string): Node[] {
  return parseInlinesAt(text, 0);
}

/** `base` is the inline depth already spent by enclosing link text: spans,
 * delimiters, and link nesting share the one <= 16 cap (a per-construct cap
 * would let composition multiply past it). */
function parseInlinesAt(text: string, base: number): Node[] {
  const chars = Array.from(text);
  const frames: Frame[] = [rootFrame()];
  let i = 0;

  while (i < chars.length) {
    const c = chars[i]!;
    if (c === "\\") {
      const n = chars[i + 1];
      if (n === "\n") {
        top(frames).children.push({ type: "hard_break" });
        i += 2;
      } else if (n !== undefined && ASCII_PUNCT.has(n)) {
        pushStr(frames, n);
        i += 2;
      } else {
        pushStr(frames, "\\");
        i += 1;
      }
    } else if (c === "`") {
      const n = runLen(chars, i, "`");
      const k = findBacktickCloser(chars, i + n, n);
      if (k !== null) {
        top(frames).children.push({ type: "code_span", text: chars.slice(i + n, k).join("") });
        i = k + n;
      } else {
        pushStr(frames, "`".repeat(n));
        i += n;
      }
    } else if (c === "[") {
      i = bracket(chars, i, false, frames, base);
    } else if (c === "!" && chars[i + 1] === "[") {
      i = bracket(chars, i + 1, true, frames, base);
    } else if (c === ":") {
      let k = i + 1;
      while (k < chars.length && isSlugChar(chars[k]!)) {
        k += 1;
      }
      const slugLen = k - (i + 1);
      if (slugLen >= 1 && slugLen <= MAX_EMOJI_SLUG_BYTES && chars[k] === ":") {
        top(frames).children.push({ type: "emoji", slug: chars.slice(i + 1, k).join("") });
        i = k + 1;
      } else {
        pushStr(frames, ":");
        i += 1;
      }
    } else if (c === "*") {
      const n = runLen(chars, i, "*");
      if (n === 1) {
        i = delimiter(chars, i, n, "em", "*", frames, base);
      } else if (n === 2) {
        i = delimiter(chars, i, n, "strong", "**", frames, base);
      } else {
        pushStr(frames, "*".repeat(n));
        i += n;
      }
    } else if (c === "~") {
      const n = runLen(chars, i, "~");
      if (n === 2) {
        i = delimiter(chars, i, n, "strike", "~~", frames, base);
      } else {
        pushStr(frames, "~".repeat(n));
        i += n;
      }
    } else {
      pushStr(frames, c);
      i += 1;
    }
  }

  // Container end: unclosed spans and delimiters revert to literal text.
  while (frames.length > 1) {
    const frame = frames.pop()!;
    top(frames).children.push(...revertFrame(frame));
  }
  const root = frames.pop()!;
  return normalize(revertDelims(root.children, root.delims));
}

function top(frames: Frame[]): Frame {
  return frames[frames.length - 1]!;
}

function pushStr(frames: Frame[], s: string): void {
  const frame = top(frames);
  // An open delimiter sits (invisibly, until it closes) at its recorded
  // index: text on its far side must not merge into text before it.
  const barrier = frame.delims.length > 0 ? frame.delims[frame.delims.length - 1]!.idx : 0;
  const children = frame.children;
  if (children.length > barrier) {
    const last = children[children.length - 1]!;
    if (last.type === "text") {
      last.value += s;
      return;
    }
  }
  children.push({ type: "text", value: s });
}

function runLen(chars: string[], i: number, c: string): number {
  let n = 0;
  while (i + n < chars.length && chars[i + n] === c) {
    n += 1;
  }
  return n;
}

function findBacktickCloser(chars: string[], from: number, n: number): number | null {
  let k = from;
  while (k < chars.length) {
    if (chars[k] === "`") {
      const m = runLen(chars, k, "`");
      if (m === n) {
        return k;
      }
      k += m;
    } else {
      k += 1;
    }
  }
  return null;
}

function totalDepth(frames: Frame[]): number {
  let depth = frames.length - 1;
  for (const f of frames) {
    depth += f.delims.length;
  }
  return depth;
}

function delimiter(
  chars: string[],
  i: number,
  n: number,
  kind: DelimKind,
  raw: string,
  frames: Frame[],
  base: number,
): number {
  const canClose = i > 0 && !isWs(chars[i - 1]!);
  const next = chars[i + n];
  const canOpen = next !== undefined && !isWs(next);
  const deep = base + totalDepth(frames) >= MAX_INLINE_DEPTH;
  const frame = top(frames);
  const innermost = frame.delims[frame.delims.length - 1];
  if (canClose && innermost !== undefined && innermost.kind === kind) {
    frame.delims.pop();
    const inner = normalize(frame.children.splice(innermost.idx));
    frame.children.push(
      kind === "em"
        ? { type: "emphasis", children: inner }
        : kind === "strong"
          ? { type: "strong", children: inner }
          : { type: "strikethrough", children: inner },
    );
  } else if (canOpen && !deep) {
    frame.delims.push({ kind, idx: frame.children.length, raw });
  } else {
    pushStr(frames, raw);
  }
  return i + n;
}

/** Handle a bracket construct starting at `chars[open]` (which is `[`).
 * `embed` means a `!` sits just before it. Returns the new position. */
function bracket(
  chars: string[],
  open: number,
  embed: boolean,
  frames: Frame[],
  base: number,
): number {
  const bang = embed ? open - 1 : open;
  const fallback = (): number => {
    pushStr(frames, chars[bang]!);
    return bang + 1;
  };

  // Find the matching `]` (balanced, escape-aware).
  let depth = 1;
  let k = open + 1;
  while (k < chars.length) {
    const c = chars[k]!;
    if (c === "\\") {
      k += 1;
    } else if (c === "[") {
      depth += 1;
    } else if (c === "]") {
      depth -= 1;
      if (depth === 0) {
        break;
      }
    }
    k += 1;
  }
  if (k >= chars.length) {
    return fallback();
  }
  const interior = chars.slice(open + 1, k).join("");

  // Link / embed: `](` with a lexable target.
  if (chars[k + 1] === "(") {
    // A link's text is one nesting level (embeds don't recurse: alt is a
    // plain string). Over the shared inline cap, fall back to literal.
    if (!embed && base + totalDepth(frames) >= MAX_INLINE_DEPTH) {
      return fallback();
    }
    const lexed = lexTarget(chars, k + 2);
    if (lexed !== null) {
      const depth = base + totalDepth(frames) + 1;
      top(frames).children.push(
        embed
          ? { type: "embed", target: lexed.target, alt: resolveEscapes(interior) }
          : { type: "link", target: lexed.target, children: parseInlinesAt(interior, depth) },
      );
      return lexed.end;
    }
    return fallback();
  }

  // Span closer: `[/name]` must name the innermost open span.
  if (interior.startsWith("/")) {
    const name = interior.slice(1);
    if (isName(name)) {
      if (embed) {
        pushStr(frames, "!"); // the ! belongs to links/embeds, not spans
      }
      if (frames.length > 1 && top(frames).name === name) {
        const frame = frames.pop()!;
        const children = normalize(revertDelims(frame.children, frame.delims));
        top(frames).children.push({ type: "span", name: frame.name, attrs: frame.attrs, children });
      } else {
        // Well-formed but mismatched or orphan: the characters back.
        pushStr(frames, `[/${name}]`);
      }
      return k + 1;
    }
    return fallback();
  }

  // Span opener: `[name ...]`, with the BBCode default-parameter idiom
  // (`[color=red]` puts `color=red` in the span's own attrs).
  const opener = parseSpanOpener(interior);
  if (opener !== null) {
    if (embed) {
      pushStr(frames, "!"); // the ! belongs to links/embeds, not spans
    }
    if (base + totalDepth(frames) >= MAX_INLINE_DEPTH) {
      pushStr(frames, chars.slice(open, k + 1).join(""));
    } else {
      frames.push({
        openerRaw: chars.slice(open, k + 1).join(""),
        name: opener.name,
        attrs: opener.attrs,
        children: [],
        delims: [],
      });
    }
    return k + 1;
  }

  return fallback();
}

function parseSpanOpener(interior: string): { name: string; attrs: Attrs } | null {
  const nlen = nameLen(interior);
  if (nlen === 0) {
    return null;
  }
  const name = interior.slice(0, nlen);
  const pairs = new Map<string, string>();
  let rest = interior.slice(nlen);
  if (rest.startsWith("=")) {
    const v = parseValue(rest.slice(1));
    if (!v.ok) {
      return null;
    }
    if (!(v.rest === "" || v.rest.startsWith(" ") || v.rest.startsWith("\t"))) {
      return null;
    }
    pairs.set(name, v.value);
    rest = v.rest;
  } else if (!(rest === "" || rest.startsWith(" ") || rest.startsWith("\t"))) {
    return null;
  }
  const parsed = parseAttrs(rest);
  if (!parsed.ok) {
    return null;
  }
  for (const [key, value] of Object.entries(parsed.attrs)) {
    if (!pairs.has(key)) {
      pairs.set(key, value);
    }
  }
  return { name, attrs: Object.fromEntries(pairs) };
}

/** Lex a link/embed target from `chars[from]`: no whitespace, balanced
 * parens, an unbalanced `)` ends it. Returns the position after `)`. */
function lexTarget(chars: string[], from: number): { target: string; end: number } | null {
  let depth = 0;
  let k = from;
  while (k < chars.length) {
    const c = chars[k]!;
    if (c === ")" && depth === 0) {
      const target = chars.slice(from, k).join("");
      if (utf8Len(target) > MAX_TARGET_BYTES) {
        return null;
      }
      return { target, end: k + 1 };
    }
    if (c === ")") {
      depth -= 1;
    } else if (c === "(") {
      depth += 1;
    } else if (isWs(c)) {
      return null;
    }
    k += 1;
  }
  return null;
}

function resolveEscapes(s: string): string {
  let out = "";
  const chars = Array.from(s);
  let i = 0;
  while (i < chars.length) {
    const c = chars[i]!;
    if (c === "\\") {
      const n = chars[i + 1];
      if (n !== undefined && ASCII_PUNCT.has(n)) {
        out += n;
        i += 2;
        continue;
      }
      out += "\\";
      if (n !== undefined) {
        out += n;
        i += 2;
        continue;
      }
      i += 1;
      continue;
    }
    out += c;
    i += 1;
  }
  return out;
}

/** An unclosed span reverts: its opener text, then its children, flattened
 * into the parent (its own unmatched delimiters reverting first). */
function revertFrame(frame: Frame): Node[] {
  const out: Node[] = [{ type: "text", value: frame.openerRaw }];
  out.push(...revertDelims(frame.children, frame.delims));
  return out;
}

function revertDelims(children: Node[], delims: Delim[]): Node[] {
  for (let d = delims.length - 1; d >= 0; d -= 1) {
    children.splice(delims[d]!.idx, 0, { type: "text", value: delims[d]!.raw });
  }
  return children;
}

/** Canonical text: adjacent literals merge, empty text nodes vanish. */
function normalize(children: Node[]): Node[] {
  const out: Node[] = [];
  for (const node of children) {
    if (node.type === "text") {
      if (node.value === "") {
        continue;
      }
      const last = out[out.length - 1];
      if (last !== undefined && last.type === "text") {
        last.value += node.value;
        continue;
      }
    }
    out.push(node);
  }
  return out;
}
